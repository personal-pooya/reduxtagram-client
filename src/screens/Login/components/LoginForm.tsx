import React, { ChangeEvent } from 'react';
import '../../../App.css';
import { Link } from 'react-router-dom';


interface Props {
  handleChange: (e: ChangeEvent) => void;
  handleSubmit: () => void;
  handleEmail: () => void;
  handlePassword: () => void
  formErrors: {email?: string, password?: string}
  failure: string | undefined;
}

class LoginForm extends React.Component<Props> { 
    render() {     
      const { formErrors, handleChange, handleEmail, handlePassword, handleSubmit, failure } = this.props;    
      const loginError = failure? failure : null;
        return (
          <>
            
            <div className="center-form">
              <h2 style={{textAlign: "center"}}>Existing User? Login here</h2>
              <h3 style={{color:"red"}}>{loginError}</h3>
              <div className="form-group">
                <label>Enter your email</label>
                <input  
                    className="form-control" 
                    type="email" 
                    name="email"
                    onBlur={handleEmail}
                    onChange={handleChange} 
                    contentEditable={true}
                >
                </input>
                <div> {formErrors ? formErrors.email : null} </div>
              </div>
              <div className="form-group">
                <label>Enter your password</label>
                <input  
                    className="form-control" 
                    type="password" 
                    name="password" 
                    onBlur={handlePassword}
                    onChange={handleChange}
                >
                </input>
                <div> {formErrors ? formErrors.password : null} </div>
              </div>
              <div className="form-group">
                  <button className="btn btn-primary"  onClick={handleSubmit}>Login</button>
                  <Link to="/signup" className="btn btn-link" style={{display: "block"}}>Click here to register a new account</Link> 
              </div>
            </div>
          </>
        );
    }
}

export default LoginForm;