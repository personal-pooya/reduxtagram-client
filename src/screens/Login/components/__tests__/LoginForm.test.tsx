import React from 'react';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-enzyme';
import LoginForm from '../LoginForm';
import toJson from "enzyme-to-json";
import { BrowserRouter as Router } from 'react-router-dom';

configure({adapter: new Adapter()});

function devNullListener(){
  return ;
}

const props = {
  handleChange: () => devNullListener,
  handleSubmit: () => devNullListener,
  handleEmail: () => devNullListener,
  handlePassword: () => devNullListener,
  formErrors: { email: '', password: ''},
  failure: ''
};

describe('Login component', () => {
  describe('should Login when credential exists in the database', () => {
    it('render loginform successfully', () => {
      /* 
        toContainReact() is a matcher function. To use it, use jest-enzyme
      */
      //const LoginComponent = shallow(<LoginForm {...props}/>);
      //expect(toDoList).toContainReact(<ul/>);
      //expect(toDoList.find('ul').length).toEqual(1)
      // const wrapper = mount(<ul/>);
      // expect(wrapper.find('ul').isEmpty()).to.equal(true);
      const LoginComponent = shallow(<LoginForm {...props}/>);
      //console.log(LoginComponent.debug())
      expect(toJson(LoginComponent)).toMatchSnapshot();
      LoginComponent.unmount()
      
    })
    it('renders only two input boxes', () => {
      const wrapper = shallow(<LoginForm {...props}/>);
     
      expect(wrapper.find("input[name='email']")).toHaveLength(1);
      expect(wrapper.find("input[name='password']")).toHaveLength(1);
      // expect(wrapper.find("input").exists()).toBe(true);
      // expect(wrapper.find("label").length).toBe(2);
      
    })
    
/* 
    it('should call handleSubmit when login button is clicked', () => {
      const wrapper = shallow(<LoginForm {...props}/>);
      console.log(wrapper.instance().props.handleSubmit)
      
      wrapper.instance().handleSubmit = jest.fn();
      const instance = wrapper.instance();
      jest.spyOn(instance, 'handleSubmit');
      wrapper.find('button').simulate('click');
    //   // expect(instance.props.handleSubmit()).toHaveBeenCalled();
    //   // wrapper.find('button').simulate('click');
     
     
    }); */
    it('should call handleSubmit when login button is clicked', () => {
     
      const handleSubmit = jest.fn();
      const wrapper = shallow(
        <LoginForm {...props} handleSubmit={handleSubmit}/>
      )
      wrapper.update();
      wrapper.find('button').simulate('click')
      expect(handleSubmit).toHaveBeenCalled();

    });

    it('should call handleSubmit when login using spy', () => {
      // wrapper.update();
    /*   const spy = jest.spyOn(loginWrapper.instance(), 'handleSubmit');
      loginWrapper.instance().forceUpdate();
      spy.mockReset();
      wrapper.setProps({}); */
      // console.log(wrapper.props().children.props.d)

      //wrapper.instance().props.handleSubmit = jest.fn();
      // const spy = jest.spyOn(wrapper.instance(), 'handleSubmit');

    // force update as the function has already been linked to some sub components,
    // so the spy function will not be used by the original one.
    // (see: https://github.com/airbnb/enzyme/issues/944)
      // wrapper.instance().forceUpdate();
      // spy.mockReset();

      
     
      /* loginWrapper.find('button').at(0).simulate("click");
      
     
      expect(spy).toHaveBeenCalled();
      spy.mockRestore();
      wrapper.unmount(); */

      // const spy = jest.spyOn(wrapper.props(), 'handleSubmit');
      // wrapper.update();
      // const mockedEvent = {target: {}};
      // wrapper.find('button').simulate('click')
      // expect(handleSubmit).toHaveBeenCalled();
     

    }); 
  });
});
