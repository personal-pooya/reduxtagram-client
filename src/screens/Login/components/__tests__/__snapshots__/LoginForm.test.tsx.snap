// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should Login when credential exists in the database render loginform successfully 1`] = `
<Fragment>
  <h2
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    Existing User? Login here
  </h2>
  <div
    className="center-form"
  >
    <h3
      style={
        Object {
          "color": "red",
        }
      }
    />
    <div
      className="form-group"
    >
      <label>
        Enter your email
      </label>
      <input
        className="form-control"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        type="email"
      />
      <div>
         
         
      </div>
    </div>
    <div
      className="form-group"
    >
      <label>
        Enter your password
      </label>
      <input
        className="form-control"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        type="password"
      />
      <div>
         
         
      </div>
    </div>
    <div
      className="form-group"
    >
      <button
        className="btn btn-primary"
        onClick={[Function]}
      >
        Login
      </button>
      <Link
        className="btn btn-link"
        to="/signup"
      >
        Register
      </Link>
    </div>
  </div>
</Fragment>
`;
